### DISCUSSION TITLE
Bug: using new storage option with createJSONStorage causes subscription to fire infinitely

### DISCUSSION THREAD

**Discussion by meypod** [pmndrs/zustand#1516] [General]
I have an app that has an intro in which it got the following hook + a dropdown to change `SELECTED_LOCALE`
```ts
  useEffect(() => {
    const unsub = settings.subscribe((state, prevState) => {
      if (state.SELECTED_LOCALE !== prevState.SELECTED_LOCALE) {
        console.log(prevState.SELECTED_LOCALE, state.SELECTED_LOCALE);
        loadLocale(state.SELECTED_LOCALE);
        I18nManager.forceRTL(isRTL);
        // allow some time for forceRTL to work
        setTimeout(restart, 200);
      }
    });

    return unsub;
  });
```
when I use `createJSONStorage` and the new `storage` option, my app data persists, but once the locale is change, which then causes listener above to fire, and my app restarts
after my app is started and the hook above is ran again, the listener is fired again, causing the app to infinitely restart

this is not happening with the deprecated `getStorage` option.

this is also the relevant code snippets:
```ts
// mmkv.ts
import {MMKV} from 'react-native-mmkv';
import {StateStorage} from 'zustand/middleware/persist';

export const storage = new MMKV();

export const zustandStorage: StateStorage = {
  setItem: (name: string, value: string) => {
    return storage.set(name, value);
  },
  getItem: (name: string) => {
    const value = storage.getString(name);
    return value ?? null;
  },
  removeItem: (name: string) => {
    return storage.delete(name);
  },
};

```
```ts
import {zustandStorage} from './mmkv';

// ..... somewhere down the file, in persist options
{
    storage: createJSONStorage(() => zustandStorage),
}
```

**Comment by dai-shi** [comment]
Thanks for reporting.
Sounds like a regression in the new storage option. Let's fix it.
Is the problem reproducible with `localStorage` with react-dom?
Or, is it the issue related only with MMKV?

**Reply by meypod** [reply]
well I have no idea how to reproduce it other than inside my project here: https://github.com/meypod/al-azan
you can checkout tag v1.10.0-rc.9 and try changing it to new option, start the app, change language and start debugging
or you can also direct me to where to look at, I can do simple debugging ;)

**Reply by dai-shi** [reply]
You can dig into `persist` code. I would probably try to reproduce the issue with localStorage or AsyncStorage first.

**Comment by ongnicolas** [comment]
Also encountering this issue with our react-native project with MMKV.

**Reply by meypod** [reply]
This has been fixed
Make sure you are using the latest version

**Reply by ongnicolas** [reply]
Latest version of zustand? What version is your MMKV and RN?

**Reply by meypod** [reply]
My versions:
Zustand 4.3.2
Rn 0.71.1
Mmkv 2.6.1

But the issue was related to zustand and was a regression (you can see it here #1517 which was fixed by #1518 ), so rn and mmkv versions doesn't matter
